alns/select/MABSelector.py

```diff
@@ -0,0 +1,216 @@
+
+    Parameters
+    ----------
+    scores
+        A list of four non-negative elements, representing the rewards when the
+        candidate solution results in a new global best (idx 0), is better than
+        the current solution (idx 1), the solution is accepted (idx 2), or
+        rejected (idx 3).
+    mab
+        A mabwiser MAB object that will be used to select the
+        (destroy, repair) operator pairs. Them arms of the ``mab`` object must
+        be generated with the static method ``make_arms``.
+    num_destroy
+        Number of destroy operators.
+    num_repair
+        Number of repair operators.
+    op_coupling
+        Optional boolean matrix that indicates coupling between destroy and
+        repair operators. Entry (i, j) is True if destroy operator i can be
+        used together with repair operator j, and False otherwise.
+    context_extractor
+        Optional function that takes a State object and returns a context
+        vector for that state that can be passed to a contextual mabwiser
+        bandit. If the MAB algorithm supports it, this context will be used to
+        help predict the next (destroy, repair) combination.
+
+    References
+    ----------
+    .. [1] Emily Strong, Bernard Kleynhans, & Serdar Kadioglu (2021).
+           MABWiser: Parallelizable Contextual Multi-armed Bandits.
+           Int. J. Artif. Intell. Tools, 30(4), 2150021:1–2150021:19.
+    """
+
+    @staticmethod
+    def make_arms(
+        num_destroy: int,
+        num_repair: int,
+        op_coupling: Optional[np.ndarray] = None,
+    ) -> List[str]:
+        """
+        Generates a list of arms for the MAB passed to MABSelector.
+
+        Any MABs passed to MABSelector must be generated with this function,
+        and with the same ``num_destroy``, ``num_repair``, and ``op_coupling``
+        parameters.
+        """
+        if num_destroy < 1 or num_repair < 1:
+            raise ValueError(
+                "num_destroy and num_repair must be greater than 0."
+            )
+        if op_coupling is not None and op_coupling.sum() == 0:
+            raise ValueError("op_coupling must have at least one True entry.")
+
+        # the set of valid operator pairs is equal to the cartesian product
+        # of destroy and repair operators, except we leave out any pairs
+        # disallowed by op_coupling
+        return [
+            f"{d_idx}_{r_idx}"
+            for d_idx, r_idx in itertools.product(
+                range(num_destroy), range(num_repair)
+            )
+            if op_coupling is None or op_coupling[d_idx, r_idx]
+        ]
+
+    def __init__(
+        self,
+        scores: List[float],
+        mab: MAB,
+        num_destroy: int,
+        num_repair: int,
+        op_coupling: Optional[np.ndarray] = None,
+        context_extractor: Optional[
+            Callable[
+                [State], Union[List[Num], np.ndarray, pd.Series, pd.DataFrame]
+            ]
+        ] = None,
+    ):
```

Comment:
I'm brainstorming a bit, but what if we construct `MAB` inside `MABSelector`, and forward any additional keyword arguments (`**kwargs`) to `MAB`? We could do something like
```python
def __init__(
  self,
  scores,
  num_destroy,
  num_repair,
  learning_policy,
  neighborhood_policy = None,
  op_coupling = None,
  **kwargs,
):
  super().__init__(...)

  # validation

  arms = self.make_arms()
  self._mab = MAB(arms, learning_policy, neighborhood_policy, **kwargs)
  # etc.
```
The benefits I see are that we keep complete control over the generation of these arms, without requiring users to first call into `MABSelector.make_arms(...)`. All they would have to do in this case is provide a learning policy, and any additional keyword arguments they want that will be passed directly to `MAB`.

Reply:
@skadio @P-bibs you know this better than I do, but: how does `MABWiser` deal with seeds? Is it deterministic if a user does not provide their own seed?

Reply:
Serdar and I went back and forth for a bit on how best to do this but had not considered this specific method. I see some upsides:
* the user doesn't need to remember to call `make_arms`, which is a bit cumbersome

And a downside:
* the parameters of the MABSelector are not evident unless you go look at an entirely different set of docs (the MABWiser docs). However, this only applies for the extra args `n_jobs`, `seed`, `backend`. `n_jobs` and `backend` will never be used in our context since parallelization is impossible (at least for now) in the ALNS context.

This leads me to wonder if we should use this approach but make `seed` an (optional) explicit argument? The only follow-on issue is if MABWiser changes their interface then we might need to update the `MABSelector` interface as well. @skadio how likely is it that another argument gets added to the `MAB` constructor at some point in the future?

Separately, I believe if no seed is provided the behavior is deterministic since [the default seed is a constant ](https://github.com/fidelity/mabwiser/blob/878873db2d072d35e26f0cc1f8d264e0319423bf/mabwiser/utils.py#L25)

Reply:
MABWiser's docs seem to also be Sphinx, so we might be able to link to them using [intersphinx](https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html). I'll try to do this sometime soon, we don't have to do so in this PR. That should make the additional arguments a bit easier to understand - a user can just click through to MABWiser's documentation from ALNS.

>This leads me to wonder if we should use this approach but make seed an (optional) explicit argument?

Good idea!

>The only follow-on issue is if MABWiser changes their interface then we might need to update the MABSelector interface as well. 

The `MABSelector` is inherently an adapter/wrapper around MABWiser. If MABWiser changes, we'll need to change in any case. I don't think we need to worry about this (too) much.